File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.py
a = int(input())
b = list(map(int, input().split()))
b.insert(0,0)
dp = [0 for _ in range(a+1)]
for i in range(1,a+1):
    for j in range(1,i+1):
        dp[i] = max(dp[i],b[j]+dp[i-j])
print(dp[a])

File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.cc
#include <stdio.h>

int max(int a, int b) {
    if (a > b) return a;
    return b;
}

int main() {
    int a;
    scanf("%d",&a);
    int b[10001];
    for (int i = 1; i <= a; i++) scanf("%d",&b[i]);
    int dp[10001] = {0};
    for (int i = 1; i <= a; i++) {
        for (int j = 1; j <= i; j++) {
            dp[i] = max(dp[i],b[j]+dp[i-j]);
        }
    }
    printf("%d",dp[a]);
    return 0;
}